planning/motion_velocity_smoother/src/motion_velocity_smoother_node.cpp

```diff
@@ -826,16 +826,31 @@ void MotionVelocitySmootherNode::applyExternalVelocityLimit(TrajectoryPoints & t
   trajectory_utils::applyMaximumVelocityLimit(
     0, traj.size(), max_velocity_with_deceleration_, traj);
 
-  const size_t closest_idx = findNearestIndexFromEgo(traj);
-
-  double dist = 0.0;
-  for (size_t idx = closest_idx; idx < traj.size() - 1; ++idx) {
-    dist += tier4_autoware_utils::calcDistance2d(traj.at(idx), traj.at(idx + 1));
+  // insert the point that has external limited velocity
+  const auto & current_pose = current_odometry_ptr_->pose.pose;
+  const size_t closest_seg_idx = motion_utils::findFirstNearestSegmentIndexWithSoftConstraints(
+    traj, current_pose, node_param_.ego_nearest_dist_threshold,
+    node_param_.ego_nearest_yaw_threshold);
+  const auto inserted_index =
+    motion_utils::insertTargetPoint(closest_seg_idx, external_velocity_limit_.dist, traj);
+  if (!inserted_index) {
+    return;
+  }
+  const double original_vel = traj.at(*inserted_index).longitudinal_velocity_mps;
+  traj.at(*inserted_index).longitudinal_velocity_mps =
+    std::min(external_velocity_limit_.velocity, original_vel);
+
+  // distance from base link to the closest trajectory point ahead of ego vehicle
+  double dist = tier4_autoware_utils::calcLongitudinalDeviation(
+    current_pose, traj.at(closest_seg_idx + 1).pose.position);
+  for (size_t idx = closest_seg_idx + 1; idx < traj.size() - 1; ++idx) {
     if (dist > external_velocity_limit_.dist) {
       trajectory_utils::applyMaximumVelocityLimit(
-        idx + 1, traj.size(), external_velocity_limit_.velocity, traj);
+        idx, traj.size(), external_velocity_limit_.velocity, traj);
       return;
     }
+
+    dist += tier4_autoware_utils::calcDistance2d(traj.at(idx), traj.at(idx + 1));
```

Comment:
It seems it is enough to apply maximum velocity from `*inserted_index` to `traj.size()`.
So, I suggest to apply maximum velocity like following:
```
trajectory_utils::applyMaximumVelocityLimit(
        *inserted_index, traj.size(), external_velocity_limit_.velocity, traj);
```
How do you think?

Reply:
Thanks! It's simpler and clearer.
I fixed in  https://github.com/autowarefoundation/autoware.universe/pull/3494/commits/87008dc1535283613cfe0a2857133fe6bc571ae0.


I moved this part to the place where the insert index was not found.
However I'm not sure why this is necessary. Is this modification correct?
```
    traj.back().longitudinal_velocity_mps = std::min(
      traj.back().longitudinal_velocity_mps, static_cast<float>(external_velocity_limit_.velocity));
```
https://github.com/autowarefoundation/autoware.universe/pull/3494/files#diff-19a3604d5e77c39fa02ac64cad147bccfa939c955d7ba1743641f4ca16d50cfdR836-R839

Reply:
My memory is a little fuzzy, but I seem to recall it was because if a deceleration point was not set, it would not slow down forever.
So, I think the modification is correct.